### DISCUSSION TITLE
Using Valet with php version installed via MacPorts

### DISCUSSION THREAD

**Discussion by davidparedes** [laravel/valet#1456] [Q&A] [answered]
Hello,

I have an old MacBook Pro running Catalina OS so Homebrew doesn't support my upgrade from php 8.0 to php 8.1. I managed to install php 8.1 using MacPorts instead of Homebrew:

````
php -v 

PHP 8.1.24 (cli) (built: Oct  2 2023 08:40:04) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.24, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.24, Copyright (c), by Zend Technologies
````

Is there anyway I can use Valet with this php version? When I try to use it Valet says it's not installed and it tries to install it again via Homebrew:
 
 ````
valet use php81 

Installing php@8.1...
[php@8.1] is not installed, installing it now via Brew... 🍻
````

Thank you!

**Comment by drbyte** [comment]
Strictly speaking, no: there's no awareness of MacPort's PHP binaries in Valet.

But, I'm exploring it ... for the same reasons you mention.

The problem is not so much having a command-line PHP binary via MacPorts ... because you can install it and just add it to your PATH. That takes only a minute to do.
The problem is actually in having php-fpm available and configurable, for Valet to recognize and link up properly.

**Comment by drbyte** [comment, marked as answer]
## Bandaging PHP versions for older Mac's: using MacPorts

TO BE CLEAR: THIS IS OFFICIALLY UNSUPPORTED AT THIS TIME. YOUR MILEAGE MAY VARY.

1. Install MacPorts via their website.

    Add the MacPorts binaries path to your PATH via your `~/.zshrc` or equivalent file so they can be used via the command-line:

    `export PATH="/opt/local/bin:/opt/local/sbin:$PATH"`

2. Install PHP and PHP-FPM for a specific version, using this command (all on one line):

```sh
    sudo port install php82 php82-calendar php82-curl php82-dba php82-exif php82-ffi php82-fpm php82-ftp \
      php82-gd php82-gettext php82-gmp php82-iconv php82-intl php82-mbstring php82-mysql php82-opcache \
      php82-openssl php82-pcntl php82-posix php82-pspell php82-soap php82-sockets php82-sodium php82-sqlite \
      php82-tidy php82-xsl php82-zip
```
   (You may have reply Yes to several prompts.) (You can add/remove related extensions depending on your needs.)

3. Configuration files for CLI:

    The base `php.ini` may not be present, so copy the dev version over.
    Make a `conf.d` folder.
    Then at the end of the `php.ini` add the directive to load all `conf.d/*.conf` files.
    Then copy the valet stub for memory limits to the `conf.d` dir.
    The following will do those for you:
```sh
    sudo cp -n /opt/local/etc/php82/php.ini-development /opt/local/etc/php82/php.ini
    sudo mkdir -p /opt/local/etc/php82/conf.d
    echo ';' | sudo tee -a /opt/local/etc/php82/php.ini
    echo 'include=/opt/local/etc/php82/conf.d/*.conf' | sudo tee -a /opt/local/etc/php82/php.ini
    echo ';' | sudo tee -a /opt/local/etc/php82/php.ini
    sudo cp -n ~/.composer/vendor/laravel/valet/cli/stubs/php-memory-limits.ini /opt/local/etc/php82/conf.d/php-memory-limits.ini
```

4. Configuration files for PHP-FPM:

    It needs both the `php-fpm.conf` file and the `valet-fpm.conf` files put in place. The following will copy the stubs over and apply configuration updates:

```sh
    sudo cp -n /opt/local/etc/php82/php-fpm.conf.default /opt/local/etc/php82/php-fpm.conf
    sudo cp -n ~/.composer/vendor/laravel/valet/cli/stubs/etc-phpfpm-valet.conf /opt/local/etc/php82/php-fpm.d/valet-fpm.conf
    sudo sed -i '' "s#VALET_USER#"$USER"#" /opt/local/etc/php82/php-fpm.d/valet-fpm.conf
    sudo sed -i '' "s#VALET_HOME_PATH#"/Users/$USER"/.config/valet#" /opt/local/etc/php82/php-fpm.d/valet-fpm.conf
    sudo sed -i '' "s#valet.sock#valet82.sock#" /opt/local/etc/php82/php-fpm.d/valet-fpm.conf
```

5. Start PHP-FPM:

    `sudo port load php82-fpm`

(You should see a new empty file named `valet82.sock` in your `~/.config/valet` directory)

6. If you haven't already, have Valet make a site config for your project, by running `valet secure` in your project directory. (This will set it up for https by creating a site-specific Nginx config file ... which we need in order to link it to a specific PHP-FPM version.)

7. Edit that Valet config in `~/.config/valet/Nginx` (ie: `foo.test` in that directory)

    Where it mentions `valet.sock` change it to `valet82.sock` instead.

    eg: `fastcgi_pass "unix:/Users/USERNAME/.config/valet/valet82.sock";`

8. Restart Nginx via Valet:

    `valet restart nginx`

9. Visit your project via `https://foo.test/` in your browser. If your foo.test app displays PHP versions, it should show the new PHP version you've linked it to via the `valet82.sock` file.
If you get a 502 error then either the foo.test Nginx config isn't pointing to the `valet82.sock` file or the `valet82.sock` file doesn't exist because that PHP-FPM version isn't running ... which would be caused by those php config files being missing, hence the elaborate config instructions above ... or just the service needs to be started with `sudo port load php82-fpm`.

There's a small possibility that you might need to run `sudo port load php82-fpm` again after a reboot. If that is a repeated problem, (sudo) edit the `/opt/local/etc/LaunchDaemons/org.macports.php82-fpm/org.macports.php82-fpm.plist` file and replace the `Disabled` key's value from `<true/>` to `<false/>`.

## NOTES:
- You can't use Valet's "PHP" commands to control any of the MacPorts PHP versions.
- If you need to also use another PHP version, repeat ALL the steps above, for that different version (replacing "82" with "83" for example)
- You CAN leave multiple FPM versions set up to run (they WILL take up RAM though), and thus you can have different sites using different PHP versions simultaneously.

If Nginx gives a 502 error, check its logs at `~/.config/valet/Log/nginx-error.log`


## PHP CLI (Command-Line)
Remember: when you call artisan commands and especially when you call composer commands, you want them to use the version of PHP your project is expecting, else you'll build outdated composer.lock files (well, that may not matter if you're not publishing your lock file to git or as part of your CI or deployment processes).

### One-time `php` commands
For one-off situations, if you need a specific PHP version on command-line, use `php82` instead of `php`.

### Set the default `php` command-line version to an installed MacPorts version
To set the default PHP version for your command-line:
- First disconnect Homebrew's PHP on the command-line. (Valet will still say "unable to determine PHP version, or PHP is not installed")

  `brew unlink php@7.4`

- Then activate the desired MacPorts version of PHP for command-line:

    `sudo port select --set php php82`

Related commands:
`port select --list php` - shows PHP versions available on your machine
`port select --summary` - shows currently linked versions of all MacPorts ports
`sudo port select --set php php82` -- sets php 8.2 to be actively linked (instead of homebrew's linked version) for CLI

`which php` and `type php` will show which php binary is gonna be called when running `php` on the CLI
`which -a php` will show all `php` binaries in all paths on your PATH, in the order in which the lookup will pick one (picks first one)

**Reply by davidparedes** [reply]
First of all, thank you so much for your awesome help! I tried it but I'm still getting a 502 error when trying to access my site. The file `valet82.sock` was generated automatically and I replaced `valet.sock` with `valet82.sock` in `mysite.test` file, but still 502 error.

These are the differences I notice with your experience:

1. The `conf.d` directory didn't exist in `/opt/local/etc/php82/` so I had to create it with `mkdir`.
2. My Valet is not linked to any php version since I uninstalled my old one and then I was not able to install any of them (8.0, 8.1, 8.2, etc.) via Homebrew. When I run `valet restart` for example I get this warning:

```
Restarting dnsmasq...

In Brew.php line 264:
                                                                         
  Homebrew PHP appears not to be linked. Please run [valet use php@X.Y]  
                                                                         
restart [<service>]
```

Do you think any of these two issues is causing the 502 error?

Thank you very much!

**Reply by drbyte** [reply]
Try calling `valet restart nginx` instead of just `valet restart`, so that it only operates on nginx. (Otherwise it restarts all of dnsmasq, php and nginx, and we don't want it touching anything to do with PHP versions in this context.)

**Reply by davidparedes** [reply]
I'm still getting "502 Bad Gateway" :(. Do you think there is any way to link Valet with an old php version? I say it just in case the issue happens due to Valet being unlinked to php. Thanks!

**Reply by drbyte** [reply]
With an older Mac it's gonna be tough to get Homebrew to install any PHP version. It might actually be helpful to check whether your backups had made a backup of `/usr/local/` (and all its subdirs), and consider restoring it so you can have a fully-functioning (old) Homebrew install .... and (almost) never upgrade it anymore.

That said, you could try to manually install a semi-recent PHP version, either 8.1 or 8.2, using `brew install php@8.1` and see what it does with its manual compile process. It'll download a ton of related libs and do all its compiling, which may take a couple hours.  But it's possible you might not get all that to work at all given the way homebrew has aggressively dumped support for older versions.

**Reply by drbyte** [reply]
If it created the valet82.sock file then the PHP-FPM service is running. 
So you'll need to look in the nginx error log file to find out why the 502 was triggered...

Look at:
`/Users/USERNAME/.config/valet/Log/nginx-error.log`

**Reply by davidparedes** [reply]
I realized I still had an old Homebrew PHP version (7.4) so I linked it, I replaced `valet.sock` with `valet82.sock` in `mysite.test` file (it had been automatically changed back to valet.sock, that's why I had to do it again now) and it works! My site is up and running again, and now with php 8.2. Thank you so much! You saved my day/week :)

**Reply by drbyte** [reply]
Yay! And good call on re-linking the 7.4. That should be fine. If at some point you "really" need 8.0 you can probably unlink homebrew and then symlink the macport binary to `php` instead.

ya, what you're working with now is a little fragile, but if your requirements don't change much you should be stable for awhile this way.

Remember when you call artisan commands and especially when you call `composer` commands, to use the php82 instead of php else you'll build outdated composer.lock files (well, that may not matter if you're not publishing your lock file to git or as part of your CI or deployment processes).

**Reply by drbyte** [reply]
Update about the CLI stuff I mentioned above:

I think if you `brew unlink php@7.4` Valet will probably still be fine because it "has" a PHP version, even if not actively symlinked.

And then run `sudo port select --set php php82` your `php` binary will be symlinked so that your CLI will be using 8.2.

Related commands:
`port select --list php` - shows PHP versions available
`port select --summary` - shows currently linked versions of all MacPorts ports
`sudo port select --set php php82` -- sets php 8.2 to be actively linked (instead of homebrew's linked version) for CLI

`which php` and `type php` will show which php binary is gonna be called when running `php` on the CLI
`which -a php` will show all `php` binaries in all paths on your PATH, in the order in which the lookup will pick one (picks first one)

**Reply by davidparedes** [reply]
Hello again,

I confirm I unlinked php 7.4 and it still works. I already had set my CLI php to 8.2, the "only" missing thing was being able to use it with Valet.

Thank you very much for your help!

**Comment by zaghadon** [comment]
I suggest using Mac ports by default to manage valet dependencies. It's way faster than Homebrew and Supports way more Devices.
